torch/csrc/jit/dynamic_dag.h

```diff
@@ -0,0 +1,523 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <type_traits>
+#include <unordered_set>
+#include <vector>
+
+#include "torch/csrc/utils/functional.h"
+
+namespace torch { namespace jit { namespace detail {
+
+// DynamicDAG is a DAG that dynamically maintains a topological order as
+// edges/vertices are added and removed.
+//
+// [Example applications]
+// - Let's say you have a DAG where each vertex is black or red. How do we
+//   merge black nodes that are directly connected by contracting the
+//   edge between them while still maintaining the DAG and a topological order?
+//   Use contractEdge().
+// - Let's say you have a DAG where each vertex is a Node* and the edges represent
+//   data dependencies. We wish to determine if adding a new Node* with certain
+//   data dependencies (or moving an existing one to use new dependencies) is valid.
+//   Use DynamicDAG::addEdge() to add the new data dependencies to the DAG:
+//   it will either find a valid reordering of the DAG's topological order or throw
+//   if the resulting DAG is invalid.
+//
+// The implementation is based off of the PK algorithm in the following paper:
+// "A Dynamic Topsort Algorithm for Directed Acyclic Graphs"
+// by David Pearce and Paul Kelly
+// https://www.doc.ic.ac.uk/~phjk/Publications/DynamicTopoSortAlg-JEA-07.pdf
+// It is summarized in [Edge addition] (see DynamicDAG<T>::addEdge)
+
+template <typename T> struct Vertex;
+template <typename T> struct DynamicDAG;
+template <typename T> using vertex_list = std::vector<Vertex<T>*>;
+template <typename T> using unique_vertex = std::unique_ptr<Vertex<T>>;
+
+enum DFSDirection {forward, backward};
+
+// Some helper functions
+template <typename T>
+struct vertex_collection {
+  static void insert(vertex_list<T>& lst, Vertex<T>* v) {
+    lst.push_back(v);
+  }
+
+  static void erase(vertex_list<T>& lst, Vertex<T>* v) {
+    lst.erase(std::find(lst.begin(), lst.end(), v));
```

Comment:
I don't think this code is correct. Those adjacency lists should effectively be sets, because we don't care about multi-edges, but what you have in here is effectively a multi-set.

Reply:
Given that we can't represent them as sorted vectors, I really think we might want to revert the changes back to a regular `unordered_set`. Sorry for all the noise 😕

Reply:
addEdges(x, y) doesn't add an edge from x to y if it already exists (DynamicDAG is used to build a dependency graph that is a simpler version of the values graph -- I should write this somewhere in a comment). Because of this, these adjacency lists should contain unique elements.

No problem, it's not hard to switch it back to `unordered_set`. I remember that one of the reasons for `vector` over `unordered_set` was that it would have better performance because of data locality and that the collections would be small. Given that, what are the benefits of using `unordered_set` instead?

Reply:
I’m ok with using vectors, but we at least should clearly indicate that this data structure has multi set semantics, and why is that ok in the context of this algorithm.